sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/TLSEvaluationStage.java

```diff
@@ -0,0 +1,276 @@
+        this.secretsClient = secretsClient;
+        this.keyPairGenerator = keyPairGenerator;
+    }
+
+    public static TLSEvaluationStage fromEnvironmentForService(
+            String serviceName,
+            String taskName) throws IOException, InvalidKeySpecException, NoSuchAlgorithmException {
+
+        SchedulerFlags flags = SchedulerFlags.fromEnv();
+
+        KeyFactory keyFactory = KeyFactory.getInstance("RSA");
+
+        PemReader pemReader = new PemReader(new StringReader(flags.getServiceAccountPrivateKeyPEM()));
+        PrivateKey privateKey = keyFactory.generatePrivate(
+                new PKCS8EncodedKeySpec(pemReader.readPemObject().getContent()));
+
+        ServiceAccountIAMTokenProvider serviceAccountIAMTokenProvider = new ServiceAccountIAMTokenProvider.Builder()
+                .setIamUrl(URLHelper.fromUnchecked(DcosConstants.IAM_AUTH_URL))
+                .setUid(flags.getServiceAccountUid())
+                .setPrivateKey((RSAPrivateKey) privateKey)
+                .build();
+        TokenProvider tokenProvider = new CachedTokenProvider(serviceAccountIAMTokenProvider);
+
+        Executor executor = Executor.newInstance(
+                new DcosHttpClientBuilder()
+                        .setTokenProvider(tokenProvider)
+                        .setRedirectStrategy(new LaxRedirectStrategy())
+                        .build());
+
+        CertificateAuthorityClient certificateAuthorityClient = new DefaultCAClient(executor);
+        SecretsClient secretsClient = new DefaultSecretsClient(executor);
+        KeyPairGenerator keyPairGenerator = KeyPairGenerator.getInstance("RSA");
+
+        return new TLSEvaluationStage(
+                serviceName, taskName, certificateAuthorityClient, secretsClient, keyPairGenerator);
+
+    }
+
+    @Override
+    public EvaluationOutcome evaluate(MesosResourcePool mesosResourcePool, PodInfoBuilder podInfoBuilder) {
+
+        // Generate new private key and encode it to PEM
+        // Generate new CSR, sign it, encode certificate to PEM
+        try {
+            KeyPair keyPair = keyPairGenerator.generateKeyPair();
+
+            X509Certificate certificate = certificateAuthorityClient.sign(generateCSR(keyPair));
+
+            String certPEM = certificateToPem(certificate);
+            String privateKeyPEM = privateKeyToPem(keyPair.getPrivate());
+
+            storeSecrets(podInfoBuilder, certPEM, privateKeyPEM);
+        } catch (Exception e) {
+            LOGGER.error("Failed to get certificate", e);
+            return EvaluationOutcome.fail(
+                    this, "Failed because of exception: %s", e);
+        }
+
+        Collection<Protos.Volume> volumes = getExecutorInfoSecretVolumes(podInfoBuilder);
+
+        // Share keys to the container
+        Optional<Protos.ExecutorInfo.Builder> executorBuilder = podInfoBuilder.getExecutorBuilder();
+        Protos.TaskInfo.Builder taskBuilder = podInfoBuilder.getTaskBuilder(taskName);
```

Comment:
nit: no need to assign `taskBuilder` here since we'll only refer to it `if (executorBuilder.isPresent())`. Move it inside the if statement.